### ISSUE TITLE
Save history or SQL snippets

### ISSUE THREAD

**Issue by hko-gh** [sqlitebrowser/sqlitebrowser#206]
It would be great to have the possibility to paste SQL code in the "Execute SQL" tab. Either having a history browser or a simple snippet manager where you can add the current code to would be very supportive. At the moment I have a text file where I paste often-used statements, which is not very comfortable.
What do you think?

**Comment by justinclift** [comment]
Hmmm yeah, both sounds useful.  A "*History Browser*" (that saves between sessions), and a "*Snippet Manager*" to organise often used SQL er... snippets, would be cool.

First question... are you able to assist in getting it done?  eg coding skills, or design skills (eg whip up some potential layout screenshots) or similar? :grin:

**Comment by sergbus** [comment]
Another useful thing might be to not close "Execute SQL" tabs when closing connection to the database. I often have a need to execute same ad-hoc query against couple databases and lose the query when closing one database and opening another. Maybe disabling the "Execute" button is enough?

**Comment by justinclift** [comment]
@sergbus That should be do-able.  We should open a brand new issue here for that though, as the code to make it happen would be tracked separately from @hko-gh's idea. :smile:

@sergbus As an thought, do you have the skill (or desire to learn) to do the implementation for this?

**Comment by sergbus** [comment]
Desire to learn. Never worked with Qt and github repos. Last C++ was first
year of college, 15 years ago.

**Comment by justinclift** [comment]
Heh, that's going to be a lot of challenge then. :wink:

**Comment by Deborah-Digges** [comment]
Hi @justinclift is anyone working on this? I think it would be a great addition to the tool! 

Do you need any volunteers for it? I'd be happy to help!

**Comment by justinclift** [comment]
@Deborah-Digges No-one's working on this yet.  You're 100% absolutely welcome to take it on.  It would be useful for lots of people. :smile:

**Comment by Deborah-Digges** [comment]
Thanks for the prompt response @justinclift!  😄 

I will definitely add myself to the mailing list. I've done some programming in C++ 11 during college for which I implemented an image editor using QT: https://github.com/Deborah-Digges/cuda-image-editor.

I will definitely need to brush up a bit before starting on this. Is there any particular part of the codebase you would suggest I start looking at?

**Comment by justinclift** [comment]
@Deborah-Digges Cool. :smile:

For starting out - depending on what kinds of things you're interested in - it might be useful to investigate how our project files work.  That's where we save meta-data about each database between sessions.  eg people's "Execute SQL" tab contents, field widths (for the Browse Data tab) and whatever else seems useful.

* https://github.com/sqlitebrowser/sqlitebrowser/blob/24267749eaa51ee5948989b8f5f3059f11379961/src/MainWindow.cpp#L2181-L2261
* https://github.com/sqlitebrowser/sqlitebrowser/blob/24267749eaa51ee5948989b8f5f3059f11379961/src/MainWindow.cpp#L2030-L2163

Am kind of thinking that looking through that might help with this "Save history" enhancement request.  Note - I'm not sure the history should be kept "per database" or if it's be useful to have the SQL history be shared across all DB's.  Kind of thinking that sharing it across all databases might be the way to go, at least initially.  Other people might have more insight though. :wink:

Looking through that project saving/opening code though should help show one way to accomplish part of it.  The next question is... where to save the history?  That I'm not sure about.  We do keep a local database in the recent versions, so maybe in that?

@MKleusberg (our lead developer) might have a better idea.  Martin would the local database be suitable for that, or would SQL history be better saved with the preference data?  I'm kind of thinking SQL history could grow pretty big over time and likely would grow to need it's own management UI bits. :wink:
